File: analytics/src/main/java/io/appmetrica/analytics/impl/proxy/validation/ReporterBarrier.java
package io.appmetrica.analytics.impl.proxy.validation;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import io.appmetrica.analytics.AdRevenue;
import io.appmetrica.analytics.IReporter;
import io.appmetrica.analytics.Revenue;
import io.appmetrica.analytics.ValidationException;
import io.appmetrica.analytics.ecommerce.ECommerceEvent;
import io.appmetrica.analytics.impl.utils.validation.NonEmptyStringValidator;
import io.appmetrica.analytics.impl.utils.validation.NonNullValidator;
import io.appmetrica.analytics.impl.utils.validation.ThrowIfFailedValidator;
import io.appmetrica.analytics.impl.utils.validation.Validator;
import io.appmetrica.analytics.profile.UserProfile;
import java.util.Map;

public class ReporterBarrier implements IReporter {

    static final Validator<String> sEventNameValidator = new ThrowIfFailedValidator<String>(
            new NonEmptyStringValidator("Event name")
    );
    static final Validator<String> sErrorMessageValidator = new ThrowIfFailedValidator<String>(
            new NonEmptyStringValidator("Error message")
    );
    static final Validator<String> sErrorIdentifierValidator = new ThrowIfFailedValidator<String>(
            new NonEmptyStringValidator("Error identifier")
    );
    static final Validator<Throwable> sUnhandledExceptionValidator = new ThrowIfFailedValidator<Throwable>(
            new NonNullValidator<Throwable>("Unhandled exception")
    );
    static final Validator<String> sStringExceptionValidator = new ThrowIfFailedValidator<String>(
            new NonEmptyStringValidator("Unhandled exception")
    );
    static final Validator<UserProfile> sUserProfileValidator = new ThrowIfFailedValidator<UserProfile>(
            new NonNullValidator<UserProfile>("User profile")
    );
    static final Validator<Revenue> sRevenueNonNullValidator = new ThrowIfFailedValidator<Revenue>(
            new NonNullValidator<Revenue>("Revenue")
    );
    static final Validator<AdRevenue> sAdRevenueNonNullValidator = new ThrowIfFailedValidator<>(
            new NonNullValidator<AdRevenue>("AdRevenue")
    );

    static final Validator<ECommerceEvent> sECommerceNonNullValidator = new ThrowIfFailedValidator<ECommerceEvent>(
            new NonNullValidator<ECommerceEvent>("ECommerceEvent")
    );

    @NonNull
    private final PluginsBarrier pluginsBarrier;

    public ReporterBarrier() {
        this(new PluginsBarrier());
    }

    @VisibleForTesting
    ReporterBarrier(@NonNull PluginsBarrier pluginsBarrier) {
        this.pluginsBarrier = pluginsBarrier;
    }

    @Override
    public void reportEvent(@NonNull String eventName) throws ValidationException {
        sEventNameValidator.validate(eventName);
    }

    @Override
    public void reportEvent(@NonNull String eventName, @Nullable String jsonValue) throws ValidationException {
        sEventNameValidator.validate(eventName);
    }

    @Override
    public void reportEvent(@NonNull String eventName, @Nullable Map<String, Object> attributes)
            throws ValidationException {
        sEventNameValidator.validate(eventName);
    }

    @Override
    public void reportError(@NonNull String message, @Nullable Throwable error) throws ValidationException {
        sErrorMessageValidator.validate(message);
    }

    @Override
    public void reportError(@NonNull String identifier, @Nullable String message) {
        sErrorIdentifierValidator.validate(identifier);
    }

    @Override
    public void reportError(
            @NonNull String identifier,
            @Nullable String message,
            @Nullable Throwable nonNullError
    ) {
        sErrorIdentifierValidator.validate(identifier);
    }

    @Override
    public void reportUnhandledException(@NonNull Throwable exception) throws ValidationException {
        sUnhandledExceptionValidator.validate(exception);
    }

    @Override
    public void resumeSession() {

    }

    @Override
    public void pauseSession() {

    }

    @Override
    public void setUserProfileID(@Nullable String profileID) {}

    @Override
    public void reportUserProfile(@NonNull UserProfile profile) throws ValidationException {
        sUserProfileValidator.validate(profile);
    }

    @Override
    public void reportRevenue(@NonNull Revenue revenue) throws ValidationException {
        sRevenueNonNullValidator.validate(revenue);
    }

    @Override
    public void reportECommerce(@NonNull ECommerceEvent event) {
        sECommerceNonNullValidator.validate(event);
    }

    @Override
    public void setStatisticsSending(boolean enabled) {

    }

    @NonNull
    @Override
    public PluginsBarrier getPluginExtension() {
        return pluginsBarrier;
    }

    @Override
    public void reportAdRevenue(@NonNull AdRevenue adRevenue) {
        sAdRevenueNonNullValidator.validate(adRevenue);
    }

    @Override
    public void putAppEnvironmentValue(@NonNull String key, @Nullable String value) {

    }

    @Override
    public void clearAppEnvironment() {

    }

    @Override
    public void sendEventsBuffer() {

    }
}

File: analytics/src/main/java/io/appmetrica/analytics/impl/proxy/validation/ReporterBarrier.kt
package io.appmetrica.analytics.impl.proxy.validation

import io.appmetrica.analytics.AdRevenue
import io.appmetrica.analytics.ModuleEvent
import io.appmetrica.analytics.ReporterConfig
import io.appmetrica.analytics.Revenue
import io.appmetrica.analytics.ecommerce.ECommerceEvent
import io.appmetrica.analytics.impl.crash.client.AllThreads
import io.appmetrica.analytics.impl.crash.client.UnhandledException
import io.appmetrica.analytics.impl.utils.validation.NonEmptyStringValidator
import io.appmetrica.analytics.impl.utils.validation.NonNullValidator
import io.appmetrica.analytics.impl.utils.validation.ThrowIfFailedValidator
import io.appmetrica.analytics.profile.UserProfile

class ReporterBarrier {

    private val eventNameValidator = ThrowIfFailedValidator(
        NonEmptyStringValidator("Event name")
    )
    private val errorMessageValidator = ThrowIfFailedValidator(
        NonEmptyStringValidator("Error message")
    )
    private val errorIdentifierValidator = ThrowIfFailedValidator(
        NonEmptyStringValidator("Error identifier")
    )
    private val unhandledExceptionValidator = ThrowIfFailedValidator(
        NonNullValidator<UnhandledException>("Unhandled exception")
    )
    private val throwableExceptionValidator = ThrowIfFailedValidator(
        NonNullValidator<Throwable>("Throwable")
    )
    private val userProfileValidator = ThrowIfFailedValidator(
        NonNullValidator<UserProfile>("User profile")
    )
    private val revenueNonNullValidator = ThrowIfFailedValidator(
        NonNullValidator<Revenue>("Revenue")
    )
    private val adRevenueNonNullValidator = ThrowIfFailedValidator(
        NonNullValidator<AdRevenue>("AdRevenue")
    )
    private val eCommerceNonNullValidator = ThrowIfFailedValidator(
        NonNullValidator<ECommerceEvent>("ECommerceEvent")
    )

    fun reportEvent(eventName: String?) {
        eventNameValidator.validate(eventName)
    }

    fun reportEvent(
        eventName: String?,
        jsonValue: String?
    ) {
        eventNameValidator.validate(eventName)
    }

    fun reportEvent(
        eventName: String?,
        attributes: Map<String?, Any?>?
    ) {
        eventNameValidator.validate(eventName)
    }

    fun reportError(
        message: String?,
        error: Throwable?
    ) {
        errorMessageValidator.validate(message)
    }

    fun reportError(
        identifier: String?,
        message: String?,
        nonNullError: Throwable?
    ) {
        errorIdentifierValidator.validate(identifier)
    }

    fun reportUnhandledException(exception: Throwable?) {
        throwableExceptionValidator.validate(exception)
    }

    fun reportUnhandledException(exception: UnhandledException?) {
        unhandledExceptionValidator.validate(exception)
    }

    fun resumeSession() {}

    fun pauseSession() {}

    fun setUserProfileID(profileID: String?) {}

    fun reportUserProfile(profile: UserProfile?) {
        userProfileValidator.validate(profile)
    }

    fun reportRevenue(revenue: Revenue?) {
        revenueNonNullValidator.validate(revenue)
    }

    fun reportECommerce(event: ECommerceEvent?) {
        eCommerceNonNullValidator.validate(event)
    }

    fun setDataSendingEnabled(enabled: Boolean?) {}

    fun reportAdRevenue(adRevenue: AdRevenue?) {
        adRevenueNonNullValidator.validate(adRevenue)
    }

    fun putAppEnvironmentValue(
        key: String?,
        value: String?
    ) {}

    fun clearAppEnvironment() {}

    fun sendEventsBuffer() {}

    fun reportAnr(allThreads: AllThreads?) {}

    fun activate(config: ReporterConfig?) {}

    fun reportEvent(moduleEvent: ModuleEvent?) {}

    fun setSessionExtra(key: String?, value: ByteArray?) {}
}
